lemminx-maven/src/main/java/org/eclipse/lemminx/maven/PlexusConfigHelper.java

```diff
@@ -0,0 +1,450 @@
+					description, defaultValue));
+			if (alias != null) {
+				parameters.add(configure(new MojoParameter(alias, getTypeDisplayName(paramType)), required, expression,
+						description, defaultValue));
+			}
+			return;
+		}
+
+		// map
+		if (Map.class.isAssignableFrom(paramClass)) {
+			// we can't do anything with maps, unfortunately
+			parameters.add(configure(new MojoParameter(name, getTypeDisplayName(paramType)).map(), required, expression,
+					description, defaultValue));
+			if (alias != null) {
+				parameters.add(configure(new MojoParameter(alias, getTypeDisplayName(paramType)).map(), required,
+						expression, description, defaultValue));
+			}
+			return;
+		}
+
+		// properties
+		if (Properties.class.isAssignableFrom(paramClass)) {
+
+			MojoParameter nested = new MojoParameter("property", "property",
+					Arrays.asList(new MojoParameter("name", "String"), new MojoParameter("value", "String")));
+
+			parameters.add(configure(new MojoParameter(name, getTypeDisplayName(paramType), nested), required,
+					expression, description, defaultValue));
+			if (alias != null) {
+				parameters.add(configure(new MojoParameter(alias, getTypeDisplayName(paramType), nested), required,
+						expression, description, defaultValue));
+			}
+		}
+
+		// collection/array
+		Type itemType = getItemType(paramType);
+		if (itemType != null) {
+			List<MojoParameter> nested = getItemParameters(realm, enclosingClass, name, itemType);
+
+			parameters.add(configure(new MojoParameter(name, getTypeDisplayName(paramType), nested), required,
+					expression, description, defaultValue));
+
+			if (alias != null) {
+				nested = getItemParameters(realm, enclosingClass, alias, itemType);
+				parameters.add(configure(new MojoParameter(alias, getTypeDisplayName(paramType), nested), required,
+						expression, description, defaultValue));
+			}
+			return;
+		}
+
+		// pojo
+		// skip classes without no-arg constructors
+		try {
+			paramClass.getConstructor();
+		} catch (NoSuchMethodException ex) {
+			return;
+		}
+
+		List<MojoParameter> nested = loadParameters(realm, paramClass);
+		parameters.add(configure(new MojoParameter(name, getTypeDisplayName(paramType), nested), required, expression,
+				description, defaultValue));
+		if (alias != null) {
+			parameters.add(configure(new MojoParameter(alias, getTypeDisplayName(paramType), nested), required,
+					expression, description, defaultValue));
+		}
+	}
+
+	public static List<MojoParameter> loadMojoParameters(PluginDescriptor descriptor, MojoDescriptor mojo,
+			MavenSession mavenSession, BuildPluginManager buildPluginManager) {
+		Class<?> clazz;
+		try {
+			clazz = mojo.getImplementationClass();
+			if (clazz == null) {
+				if (descriptor.getClassRealm() == null) {
+					// from
+					// org.eclipse.m2e.editor.mojo.MojoParameterMetadataProvider.getPluginDescriptor(ArtifactKey,
+					// IMavenExecutionContext, IProgressMonitor)
+					// TODO: Maybe this should occur in MavenPlugin.java?
```

Comment:
This is the only real difference in this class from m2e

Reply:
This TODO isn't possible as the MavenPlugin class has no PluginDescriptor. However, MavenPluginUtils does.